### PR TITLE
Fix performance regressions

### PR DESCRIPTION
**Pull request by maleadt**
... sadly, by reverting https://github.com/JuliaGPU/CUDA.jl/pull/1153, because of:

```julia
julia> UInt32(0) ÷ typemax(Int)-1
0xffffffffffffffff
```

The fact that this results in a UInt64 easily causes conversions to Int64, which are checked. I would have expected the UInt32 source type to preserve.

EDIT: now working with Int32 instead.

### CODE REVIEW

**Comment by maleadt** [comment]
Another issue, due to #1110:

```
julia> function maybe_throw()
           Int8(threadIdx().x)
           return
       end
maybe_throw (generic function with 1 method)
```

```
julia> CUDA.code_ptx(maybe_throw, Tuple{})
.visible .func julia_maybe_throw_5413()
{
        .reg .pred      %p<2>;
        .reg .b32       %r<2>;

// %bb.0:                               // %top
        mov.u32         %r1, %tid.x;
        setp.lt.u32     %p1, %r1, 127;
        @%p1 bra        LBB0_2;
        bra.uni         LBB0_1;
LBB0_2:                                 // %L27
        ret;
LBB0_1:                                 // %L23
        { // callseq 17, 0
        .reg .b32 temp_param_reg;
        call.uni 
        julia__throw_inexacterror_5420, 
        (
        );
        } // callseq 17
                                        // -- End function
}
```
```
julia> CUDA.code_ptx(maybe_throw, Tuple{}; kernel=true)
.visible .entry _Z22julia_maybe_throw_5423(
        .param .align 8 .b8 _Z22julia_maybe_throw_5423_param_0[8]
)
{
        .local .align 8 .b8     __local_depot0[8];
        .reg .b64       %SP;
        .reg .b64       %SPL;
        .reg .pred      %p<2>;
        .reg .b32       %r<2>;
        .reg .b64       %rd<5>;

// %bb.0:                               // %entry
        mov.u64         %SPL, __local_depot0;
        cvta.local.u64  %SP, %SPL;
        ld.param.u64    %rd1, [_Z22julia_maybe_throw_5423_param_0];
        add.u64         %rd2, %SP, 0;
        add.u64         %rd3, %SPL, 0;
        st.local.u64    [%rd3], %rd1;
        mov.u32         %r1, %tid.x;
        setp.lt.u32     %p1, %r1, 127;
        @%p1 bra        LBB0_2;
        bra.uni         LBB0_1;
LBB0_2:                                 // %L27
        ret;
LBB0_1:                                 // %L23
        { // callseq 18, 0
        .reg .b32 temp_param_reg;
        .param .b64 param0;
        st.param.b64    [param0+0], %rd2;
        call.uni 
        julia__throw_inexacterror_5430, 
        (
        param0
        );
        } // callseq 18
                                        // -- End function
}
```

Notice the local storage. Ideally we'd just pass the parameter directly.

**Comment by maleadt** [comment]
This looks like an NVPTX issue:

```cuda
struct KernelState {
    int* exception_flag;
};

__device__ __attribute__((noinline)) void julia__throw_inexacterror(const KernelState * state) {
    *state->exception_flag = 1;
    return;
}

__global__ void julia_maybe_throw(const KernelState state) {
    int i = threadIdx.x;
    if (i)
        julia__throw_inexacterror(&state);
    return;
}
```

```
$ nvcc -arch=sm_70 wip.cu -ptx -o -                                                                                                                                     *[master] 
//
// Generated by NVIDIA NVVM Compiler
//
// Compiler Build ID: CL-30300941
// Cuda compilation tools, release 11.4, V11.4.120
// Based on NVVM 7.0.1
//

.version 7.4
.target sm_70
.address_size 64


.func _Z25julia__throw_inexacterrorPK11KernelState(
        .param .b64 _Z25julia__throw_inexacterrorPK11KernelState_param_0
)
{
        .reg .b32       %r<2>;
        .reg .b64       %rd<3>;


        ld.param.u64    %rd1, [_Z25julia__throw_inexacterrorPK11KernelState_param_0];
        cvta.to.global.u64      %rd2, %rd1;
        mov.u32         %r1, 1;
        st.global.u32   [%rd2], %r1;
        ret;

}
        // .globl       _Z17julia_maybe_throw11KernelState
.visible .entry _Z17julia_maybe_throw11KernelState(
        .param .align 8 .b8 _Z17julia_maybe_throw11KernelState_param_0[8]
)
{
        .reg .pred      %p<2>;
        .reg .b32       %r<2>;
        .reg .b64       %rd<2>;


        ld.param.u64    %rd1, [_Z17julia_maybe_throw11KernelState_param_0];
        mov.u32         %r1, %tid.x;
        setp.eq.s32     %p1, %r1, 0;
        @%p1 bra        $L__BB1_2;

        { // callseq 0, 0
        .reg .b32 temp_param_reg;
        .param .b64 param0;
        st.param.b64    [param0+0], %rd1;
        call.uni 
        _Z25julia__throw_inexacterrorPK11KernelState, 
        (
        param0
        );
        } // callseq 0

$L__BB1_2:
        ret;

}
```

vs

```
$ clang -O3 -S  --cuda-device-only --cuda-gpu-arch=sm_70 -o - wip.cu
//
// Generated by LLVM NVPTX Back-End
//

.version 7.0
.target sm_70
.address_size 64

        // .globl       _Z25julia__throw_inexacterrorPK11KernelState

.visible .func _Z25julia__throw_inexacterrorPK11KernelState(
        .param .b64 _Z25julia__throw_inexacterrorPK11KernelState_param_0
)
{
        .reg .b32       %r<2>;
        .reg .b64       %rd<3>;

        ld.param.u64    %rd1, [_Z25julia__throw_inexacterrorPK11KernelState_param_0];
        ld.u64  %rd2, [%rd1];
        mov.u32         %r1, 1;
        st.u32  [%rd2], %r1;
        ret;

}
        // .globl       _Z17julia_maybe_throw11KernelState
.visible .entry _Z17julia_maybe_throw11KernelState(
        .param .align 8 .b8 _Z17julia_maybe_throw11KernelState_param_0[8]
)
{
        .local .align 8 .b8     __local_depot1[8];
        .reg .b64       %SP;
        .reg .b64       %SPL;
        .reg .pred      %p<2>;
        .reg .b32       %r<2>;
        .reg .b64       %rd<5>;

        mov.u64         %SPL, __local_depot1;
        cvta.local.u64  %SP, %SPL;
        add.u64         %rd1, %SP, 0;
        add.u64         %rd2, %SPL, 0;
        ld.param.u64    %rd3, [_Z17julia_maybe_throw11KernelState_param_0];
        st.local.u64    [%rd2], %rd3;
        mov.u32         %r1, %tid.x;
        setp.eq.s32     %p1, %r1, 0;
        @%p1 bra        LBB1_2;
        { // callseq 0, 0
        .reg .b32 temp_param_reg;
        .param .b64 param0;
        st.param.b64    [param0+0], %rd1;
        call.uni 
        _Z25julia__throw_inexacterrorPK11KernelState, 
        (
        param0
        );
        } // callseq 0
LBB1_2:
        ret;

}
```

**Comment by maleadt** [comment]
@Artem-B, https://github.com/JuliaGPU/CUDA.jl/pull/1167#issuecomment-928954463 still reproduces on LLVM#main (0446f1299f6be9fd35bc5f458c78b34dca3105f6). Since you mentioned in LLVM bug #49877 that there's been back-end improvements to avoid these local copies, any thoughts on this? What would it take to improve the generated code here? Or to work around the issue (i.e. tell LLVM that uses _will_be read-only)?

For some backstory here, we're considering passing an additional kernel state object to every kernel with session- or stream-dependent information (e.g. a mapped pointer where to write thrown exceptions to, or the location of memory for a bump allocator). That object is passed to kernels by-value (to enable fast access by each thread), and to called device functions by-reference (~so that it could also be used to contain some state~ that will never work without a copy as parameter space is read-only). But it seems like that design significantly impacts generate code by introducing a local copy.

**Comment by maleadt** [comment]
It doesn't even need to be a conditional use.

```cuda
struct KernelState {
    int* exception_flag;
};

__device__ __attribute__((noinline)) void julia__throw_inexacterror(const KernelState state) {
    *state.exception_flag = 1;
    return;
}

__global__ void julia_maybe_throw(const KernelState state) {
    julia__throw_inexacterror(state);
    return;
}
```

```llvm
target datalayout = "e-i64:64-i128:128-v16:16-v32:32-n16:32:64"
target triple = "nvptx64-nvidia-cuda"

%KernelState = type { i32* }

define dso_local void @escape(%KernelState* %state) {
entry:
  ret void
}

define dso_local void @kernel(%KernelState* nocapture readonly byval(%KernelState) align 8 %state) {
entry:
  %copy = alloca %KernelState, align 8
  %0 = bitcast %KernelState* %copy to i64*
  %1 = bitcast %KernelState* %state to i64*
  %2 = load i64, i64* %1
  store i64 %2, i64* %0

  call void @escape(%KernelState* %copy)
  ret void
}
```

```ptx
.visible .func kernel(
        .param .align 8 .b8 kernel_param_0[8]
)                                       // @kernel
{
        .local .align 8 .b8     __local_depot1[8];
        .reg .b64       %SP;
        .reg .b64       %SPL;
        .reg .b64       %rd<4>;

// %bb.0:                               // %entry
        mov.u64         %SPL, __local_depot1;
        cvta.local.u64  %SP, %SPL;
        add.u64         %rd1, %SP, 0;
        add.u64         %rd2, %SPL, 0;
        ld.param.u64    %rd3, [kernel_param_0];
        st.local.u64    [%rd2], %rd3;
        { // callseq 0, 0
        .reg .b32 temp_param_reg;
        .param .b64 param0;
        st.param.b64    [param0+0], %rd1;
        call.uni 
        escape, 
        (
        param0
        );
        } // callseq 0
        ret;
                                        // -- End function
}
```

I guess I just better pass the state by value, so that LLVM doesn't need to figure this out.

**Comment by Artem-B** [comment]
Let me rephrase -- local copies will be eliminated if compiler can prove that the variable address has not been accessed and that variable has not been modified. Previously we've created local copies for all parameters and they stuck around most of the time, sometimes even when they were never accessed. 

The ability to 'prove' that the address is only taken for read operations is pretty limited. It only chacks for pointer arithmetics followed by a read. Anything else will be treated as if the pointer may escape. Explicitly taking the pointer and letting it escape on the source code level will guarantee existence of a local copy. PTX does not give us much wiggle room there when it comes to accessing the parameters.

Unfortunately, AFAICT there's no way to tell clang to pass structs by value for non-kernel functions. If you look at the IR for `julia__throw_inexacterror`, it does pass the struct by pointer and that's what forces having a local copy.
https://godbolt.org/z/vEYerEqxz

If you were to change `escape` in your IR example to make 'state' `nocapture readonly byval`, it could help.

**Comment by maleadt** [comment]
I see, thanks! Marking the state `nocapture readonly byval` doesn't seem to work though... I guess I should look at changing our calling convention to be pass-by-value, but that's pretty invasive.

EDIT: `byref` doesn't help either.

**Comment by Artem-B** [comment]
If you can compile your code with `--debug-only=nvptx-lower-args` passed to LLVM, it should print what exactly blocks direct access to the parameter.
If you could send me the IR before the `nvptx-lower-args` pass, I can take a closer look at what's going on.

**Comment by Artem-B** [comment]
I think I have and idea what might be going on: https://godbolt.org/z/1bKzor3jr

Passing parameter of one function to another as a `byval` argument is currently not on the 'safe' list and therefore blocks elimination of the local copy. I think this specific case should be easy to fix.

**Comment by Artem-B** [comment]
I was too optimistic. Allowing byval args to be passed w/o a local copy turned out to be more interesting than I thought.

The problem is that passing byval aggregates still technically uses a pointer and regular functions operate on the pointers in generic AS. At least, if you write the code in C++. There's no easy way to express that the byval value that we'll need to read from its nominally generic pointer argument should actually be read from param AS. And if we explicitly addrspacecast that pointer to param AS, we'll need to change the signature of the function to change AS of the pointer arguments depending on what's passed at each callsite. I.e. in some cases we may pass stuff that lives in param AS, in others -- something with generic pointer. In order to make it work, we'll need to be able to automatically create specialized function overloads for particular AS of the pointer arguments. That complicates things quite a bit. It's unlikely that this will get implemented any time soon.

We may be able to deal with it by adding a special kind of addrspacecast (or an opaque intrinsic) that would keep the param space pointer around and which serve as a special hint for the lowering code telling NVPTX back-end that particular byval parameter should actually be loaded from the preserver param-space pointer. That's a bit of a kludge, but it may be sufficient. The downside is that this special case will likely get in the way of optimizations, so it's not clear whether it will be a net benefit.

If you generate all IR yourself and are not constrained by the calling conventions, you can actually pass the aggregates as values and avoid pointer AS issues altogether. E.g: https://godbolt.org/z/n44cY9Tcn
Interoperating with C++ code might get tricky, though.

Another thing to consider would be to inline the functions, if possible. If we don't have to pass kernel arguments further down, it makes the local copy a bit more likely to be unnecessary.

**Comment by maleadt** [comment]
Thanks for looking into this, @Artem-B! Passing the kernel state object by value is exactly what I opted to do (for now) in https://github.com/JuliaGPU/GPUCompiler.jl/pull/251, and that seems to fix the performance regression noted here indeed. Doing that for all immutable structs will require a change to the Julia calling convention, which is more work, but might be easier than fixing LLVM/NVPTX's treatment of `byval`.
